core/include/scipp/core/values_and_variances_strong_types.h

```diff
@@ -0,0 +1,202 @@
+
+template <class T, class... Args>
+constexpr bool is_type_in_pack_v =
+    std::disjunction<std::is_same<T, std::decay_t<Args>>...>::value;
+
+template <class Tag, class... Args>
+constexpr bool is_tag_in_pack_v =
+    std::disjunction<has_tag<Tag, Args>...>::value;
+
+template <class T, template <class T1, class T2> class Cond, class... Args>
+class Indexer {
+  template <std::size_t... IS>
+  static constexpr auto indexOfCorresponding_impl(std::index_sequence<IS...>) {
+    return ((Cond<T, Args>::value * IS) + ...);
+  }
+
+public:
+  static constexpr auto indexOfCorresponding() {
+    return indexOfCorresponding_impl(
+        std::make_index_sequence<sizeof...(Args)>{});
+  }
+};
+
+template <class T, class Tuple, std::size_t... I>
+constexpr T make_move_from_tuple_impl(Tuple &&t, std::index_sequence<I...>) {
+  return T(std::move(std::get<I>(t))...);
+}
+
+template <class T, class Tuple> constexpr T make_move_from_tuple(Tuple &&t) {
+  return detail::make_move_from_tuple_impl<T>(
+      std::forward<Tuple>(t),
+      std::make_index_sequence<
+          std::tuple_size_v<std::remove_reference_t<Tuple>>>{});
+}
+
+template <class VarT, class... Ts> class ConstructorArgumentsMatcher {
+public:
+  template <class... NonDataTypes> constexpr static void checkArgTypesValid() {
+    constexpr int nonDataTypesCount =
+        (is_type_in_pack_v<NonDataTypes, Ts...> + ...);
+    constexpr bool hasVal = is_tag_in_pack_v<ValuesTag, Ts...>;
+    constexpr bool hasVar = is_tag_in_pack_v<VariancesTag, Ts...>;
+    static_assert(
+        nonDataTypesCount + hasVal + hasVar == sizeof...(Ts),
+        "Arguments: units::Unit, Shape, Dims, Values and Variences could only "
+        "be used. Example: Variable(dtype<float>, units::Unit(units::kg), "
+        "Shape{1, 2}, Dims{Dim::X, Dim::Y}, Values({3, 4}))");
+  }
+
+  template <class ElemT, class... NonDataTypes>
+  static VarT construct(Ts &&... ts) {
+    auto tp = std::make_tuple(std::forward<Ts>(ts)...);
+    auto val = std::move(extractTagged<ValuesTag, Ts...>(tp).tuple);
+    auto var = std::move(extractTagged<VariancesTag, Ts...>(tp).tuple);
+    return VarT::template createVariable<ElemT>(
+        std::forward<NonDataTypes>(extractArgs<NonDataTypes, Ts...>(tp))...,
+        std::move(val), std::move(var));
+  }
+
+private:
+  template <class T, class... Args>
+  static decltype(auto) extractArgs(std::tuple<Args...> &tp) {
+    if constexpr (!is_type_in_pack_v<T, Ts...>)
+      return T{};
+    else {
+      constexpr auto index =
+          Indexer<T, std::is_same, Args...>::indexOfCorresponding();
+      return std::get<index>(tp);
+    }
+  }
+
+  template <class Tag, class... Args>
+  static decltype(auto) extractTagged(std::tuple<Args...> &tp) {
+    if constexpr (!is_tag_in_pack_v<Tag, Ts...>)
+      return TaggedTuple<Tag>();
+    else {
+      constexpr auto index =
+          Indexer<Tag, has_tag, Args...>::indexOfCorresponding();
+      return std::get<index>(tp);
+    }
```

Comment:
Can you remove the current tag mechanism (which requires quite a few helpers) if you make the tag a base class of `Values` and `Variances`? Then you can just use `std::is_base_of` here?